Bluetooth: GATT: Persist Service Changed data

fnde-ot opened the pull request:
Fixes #19284.

Add support for persisted Service Changed data, to fix the case of a
paired device not reconnecting before a reboot and thus not receiving
SC indication. It also enables support for GATT database being changed
during a firmware update.

- Add a BT_GATT_SERVICE_CHANGED config option to enable SC support. Both
BT_GATT_CACHING and BT_GATT_DYNAMIC_DB depend on this new parameter.

- Move Service Changed data outside of the CCC struct and make it
persistent by adding support for a "bt/sc/..." setting.

Signed-off-by: François Delawarde <fnde@oticon.com>

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,136 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+
+		u8_to_dec(id_str, sizeof(id_str), cfg->id);
+		bt_settings_encode_key(key, sizeof(key), "sc",
+				       (bt_addr_le_t *)&cfg->peer, id_str);
```
The typecast seems unnecessary since cfg->peer is already a bt_addr_le_t

fnde-ot replied:
Fixed in latest push

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,136 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+				       (bt_addr_le_t *)&cfg->peer, id_str);
+	} else {
+		bt_settings_encode_key(key, sizeof(key), "sc",
+				       (bt_addr_le_t *)&cfg->peer, NULL);
```
Same here

fnde-ot replied:
Fixed in latest push

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,136 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+	if (!sc->start && !sc->end) {
+		/* Empty cfg data, just remove the setting */
+		data = NULL;
+		len = 0;
```
Use `settings_delete()` instead to make the code more readable. I'd suggest something like:
```
if (!sc->start && !sc->end) {
        err = settings_delete(key);
} else {
        err = settings_save_one(key, cfg->data, sizeof(cfg->data));
}

if (err) {
        ...
}
```

fnde-ot replied:
Fixed in latest push

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3822,6 +4003,70 @@ static int ccc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		/* Clear configuration */
+		memset(cfg->data, 0, sizeof(cfg->data));
+
+		BT_DBG("Removed SC for %s", bt_addr_le_str(&addr));
```
Shouldn't you also be doing this:
```
cfg->id = BT_ID_DEFAULT;
 bt_addr_le_copy(&cfg->peer, BT_ADDR_LE_ANY);
```

At least you seem to be doing that in the `sc_clear()` function, and this situation is essentially the same.

fnde-ot replied:
Fixed in latest push, I changed it to use sc_clear, as we also need to persist the data.

jhedberg replied:
Hmm? I don't see that change. Also, if you did do it I think it wouldn't be correct, since here you're reading data from flash, and trying to write something back to flash by calling `sc_clear()` would create a sort of loopback.

fnde-ot replied:
Sorry I misread the line number in your comment. I thought you were talking about the memset(cfg->data, 0, sizeof(cfg->data)); in the sc_restore function, where I know call sc_clear.

Just pushed a fix to this one as well in sc_set(...).

jhedberg reviewed (changes requested):
A few initial comments, I'll try to do a more thorough review later

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1464,40 +1602,54 @@ static int gatt_indicate(struct bt_conn *conn, u16_t handle,
 
-	memcpy(&data, params->data, params->len);
+	data.start = sys_le16_to_cpu(start);
+	data.end = sys_le16_to_cpu(end);
```
I'd suggest you do the byte-order conversion *before* calling this function (or inline in the function call parameters). Unlike e.g. Linux we don't have byte-order specific integer types like `le16` and `be16` so the risk of loosing track is greater when we pass around values. Because of that I'd recommend that we try to stick to protocol byte order *only* for the values inside packed structs or protocol buffers (byte arrays). Whenever the values are outside of those they should always be in host byte order.

fnde-ot replied:
I agree, it was just the way it was done before and I wanted to avoid adding too many unrelated changes. I will fix and push again soon!

fnde-ot replied:
Pushed!

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,129 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+struct gatt_sc_cfg {
+	u8_t                    id;
+	bt_addr_le_t		peer;
+	u8_t			data[4] __aligned(4);
```
I guess this is only going to be use for start and end range so I don't think we need to keep using an opaque storage.

fnde-ot replied:
Good idea, fixed in latest push.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -786,31 +913,35 @@ static bool update_range(u16_t *start, u16_t *end, u16_t new_start,
+#if defined(CONFIG_BT_GATT_CACHING)
 static void db_changed(void)
 {
-#if defined(CONFIG_BT_GATT_CACHING)
```
I rather have it defined here instead of around each time we call db_changed, not sure why you have decided to change it?

fnde-ot replied:
I thought it would be best to just ifdef out everything that had to do with a specific option when not selected, makes it easier to understand for me, but that's a matter of opinion. :)

Reverted in latest push.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1527,11 +1679,16 @@ static u8_t notify_cb(const struct bt_gatt_attr *attr, void *user_data)
+
+				sc = (struct sc_data *)data->ind_params->data;
+				sc_save(cfg->id, &cfg->peer,
+					sys_le16_to_cpu(sc->start),
```
So this is suppose to store in LE format always? Afaik we use host byte order for settings.

fnde-ot replied:
We actually store in host byte order, that is why this line converts the data from LE to CPU before calling sc_save.

Vudentz replied:
Hmm but then why the ind_params->data is carrying that in the over the air order? unpacked structs should always be in host byte order to avoid confusion.

fnde-ot replied:
Indeed, it seems that gatt_indicate should be the one to perform that sys_cpu_to_le16 right before calling gatt_send.
Not sure if I should change this in this PR as it would break the exposed API (bt_gatt_indicate) that currently expects le16 data as input.
Maybe this should be a github issue or another PR, what do you say?

Vudentz replied:
Right, well I guess we should mark sc_data as packed then since it is indeed the data format to be sent over the air. Note that bt_gatt_indicate data byte order is not defined by the spec, this is up to the profile/service to define so we can't make any assumptions on byte order.

fnde-ot replied:
> Well perhaps it not that easy to split the changes now that they are put together but logically there are independent changes, there is a dependency on the service changed Kconfig by the SC data persistency but that is about it. The advantage here is that the history is cleaner, we can bysect the changes independently, for instance you can have a clear explanation why BT_GATT_SERVICE_CHANGED may be needed even with dynamic db support disabled.

Makes sense, I'll give it a go then!

fnde-ot replied:
Yes that must be the reason it was like that! Added __packed to sc_data in the latest push.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3682,7 +3863,21 @@ int bt_gatt_clear(u8_t id, const bt_addr_le_t *addr)
+	}
+#endif /* CONFIG_BT_GATT_SERVICE_CHANGED */
+
+#if defined(CONFIG_BT_GATT_CACHING)
```
Im seeing this right or there 2 times the same code in here?

fnde-ot replied:
Done in last push!

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3904,6 +4177,9 @@ static int db_hash_commit(void)
 
 	BT_HEXDUMP_DBG(db_hash, sizeof(db_hash), "New Hash: ");
 
+	/* Indicate Service Changed */
```
We should probably have a comment here why we are resetting the range, this is most likely going to happen after a firmware update where the database changes causing the hash to be different therefore all service changed data needs to be resetted.

fnde-ot replied:
Done in last push!

Vudentz commented:
It might be better to split the changes that introduces the SERVICE_CHANGED Kconfig to the rest of the code that is introducing the storage for the SC.

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1128,6 +1265,7 @@ void bt_gatt_foreach_attr_type(u16_t start_handle, u16_t end_handle,
 		}
 	}
 
+#if defined(CONFIG_BT_GATT_DYNAMIC_DB)
```
Done in latest push! I also put the ifdef 'inside' the function block so we can avoid ifdefing the call to that function (I understand you prefer code with less ifdefs).

fnde-ot commented:
> It might be better to split the changes that introduces the SERVICE_CHANGED Kconfig to the rest of the code that is introducing the storage for the SC.

I am not sure if I can do this easily in two steps as both changes are strongly related.
Tell me if you think it is necessary and I can give it a go!

Vudentz commented:
> I am not sure if I can do this easily in two steps as both changes are strongly related.
> Tell me if you think it is necessary and I can give it a go!

Well perhaps it not that easy to split the changes now that they are put together but logically there are independent changes, there is a dependency on the service changed Kconfig by the SC data persistency but that is about it. The advantage here is that the history is cleaner, we can bysect the changes independently, for instance you can have a clear explanation why BT_GATT_SERVICE_CHANGED may be needed even with dynamic db support disabled.

fnde-ot commented:
> > I am not sure if I can do this easily in two steps as both changes are strongly related.
> > Tell me if you think it is necessary and I can give it a go!
> 
> Well perhaps it not that easy to split the changes now that they are put together but logically there are independent changes, there is a dependency on the service changed Kconfig by the SC data persistency but that is about it. The advantage here is that the history is cleaner, we can bysect the changes independently, for instance you can have a clear explanation why BT_GATT_SERVICE_CHANGED may be needed even with dynamic db support disabled.

Makes sense, I'll give it a go then!

fnde-ot committed:
bluetooth: host: Introduce a config option for Service Changed

The Service Changed characteristic support should also work when the
GATT database has been modified after reboot (firmware update scenario)

This commit introduces a BT_GATT_SERVICE_CHANGED config option that is
independent from BT_GATT_CACHING and BT_GATT_DYNAMIC_DB.

Signed-off-by: François Delawarde <fnde@oticon.com>

fnde-ot committed:
bluetooth: host: Persist Service Changed data

Add support for persisted Service Changed data, to fix the case of a
paired device not reconnecting before a reboot and thus not receiving
SC indication. It also enables support for GATT database being changed
during a firmware update.

Move Service Changed data outside of the CCC struct and make it
persistent by adding support for a bt/sc/... setting.

Signed-off-by: François Delawarde <fnde@oticon.com>

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1813,11 +1972,11 @@ static u8_t update_ccc(const struct bt_gatt_attr *attr, void *user_data)
-#if defined(CONFIG_BT_GATT_DYNAMIC_DB)
-			if (ccc->cfg_changed == sc_ccc_cfg_changed) {
-				sc_restore(&ccc->cfg[i]);
+			if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED)) {
```
Merge the nested if-statements with && ?

fnde-ot replied:
Done in latest push!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -493,19 +604,22 @@ static void remove_cf_cfg(struct bt_conn *conn)
 	BT_GATT_CHARACTERISTIC(BT_UUID_GATT_SC, BT_GATT_CHRC_INDICATE,
 			       BT_GATT_PERM_NONE, NULL, NULL, NULL),
-	BT_GATT_CCC(sc_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE),
+	BT_GATT_CCC_MANAGED((&(struct _bt_gatt_ccc)
```
Maybe make this a named struct, and then we can compare ccc with that struct instead of checking if the cfg_changed function is sc_ccc_cfg_changed. (Line 1680)

fnde-ot replied:
Done in latest push!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -674,14 +790,14 @@ static void sc_process(struct k_work *work)
 	sc->params.len = sizeof(sc_range);
 
-	if (bt_gatt_indicate(NULL, &sc->params)) {
+	if (bt_gatt_indicate(sc->conn, &sc->params)) {
```
gatt_sc is a singleton. Should we maybe remove the CONTAINER_OF in this function, and use gatt_sc ?

fnde-ot replied:
Done in latest push!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1464,40 +1606,47 @@ static int gatt_indicate(struct bt_conn *conn, u16_t handle,
-	struct sc_data data;
-	struct sc_data *stored;
+	struct gatt_sc_cfg *cfg;
+	bool modified;
```
= false;

fnde-ot replied:
Done in latest push!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1464,40 +1606,47 @@ static int gatt_indicate(struct bt_conn *conn, u16_t handle,
 	/* Check if there is any change stored */
-	if (!stored->start && !stored->end) {
-		*stored = data;
+	if (!cfg->data.start && !cfg->data.end) {
```
Same readability comment here. It's easier to read if using the same expression.

fnde-ot replied:
Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,122 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+	}
+#endif /* CONFIG_BT_SETTINGS */
+
+	cfg->id = BT_ID_DEFAULT;
```
Do you mean I should use 0 instead of BT_ID_DEFAULT for this assignment (it's a u8_t so memset might be overkill)? The next line already uses memset (bt_addr_le_copy) to set the 'peer'.

joerchan replied:
No I meant. memset(cfg, 0, sizeof(*cfg));

joerchan replied:
Just a suggestion to memset the cfg struct. As all the assignments are to 0.

fnde-ot replied:
Yes of course, I agree it makes it simpler... Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3671,6 +3830,20 @@ static int bt_gatt_clear_cf(u8_t id, const bt_addr_le_t *addr)
+
+}
+
+static int bt_gatt_clear_sc(u8_t id, const bt_addr_le_t *addr)
```
I just tried to follow naming conventions from that file to make it more readable, or easier to change in a future cleanup, as we already have:
static int bt_gatt_clear_ccc(...);
static int bt_gatt_clear_cf(...);

(I thought about renaming the above two functions but I'd rather not add too much to this PR)
Would you still prefer for me to rename it?

joerchan replied:
Yes, I noticed that they other functions where named the same way. Almost all functions in this file name `bt_gatt_*` is non-static. So I would at least rename the ones you added.

It's ok to leave the others, as they are not part of your changes but if you decide to rename them you could do it in a new commit, so that it is easy to separate the changes.

fnde-ot replied:
Pushed in last commit!

fnde-ot commented on include/bluetooth/gatt.h:
```diff
@@ -563,7 +563,6 @@ struct bt_gatt_ccc_cfg {
 	u8_t                    id;
 	bt_addr_le_t		peer;
 	u16_t			value;
-	u8_t			data[4] __aligned(4);
```
Done in latest push!

joerchan replied:
@Vudentz @jhedberg Comment to this? I'm not 100% sure

jhedberg replied:
I think we should just remove it. Wasn't this essentially an internal API, i.e. not something directly accessed by the app?

joerchan replied:
It is enclosed by struct _bt_gatt_ccc. But I don't know.
At least I couldn't find any reference to it in our SDK, so it looks fine to remove it to me.
Maybe @Vudentz can confirm if this is considered public API or not?

fnde-ot replied:
It is technically in a public header, so existing apps might use it for some purpose other than SC data storage. However, it is also very unlikely that anyone currently uses this.

Vudentz replied:
Well it is kind hidden nowadays that BT_GATT_CCC declares it internally, in any case it was never intended to be public we just to advanced of the padding to store the SC data so it should be fine to remove it.

fnde-ot replied:
Deal! I just removed it in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,122 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+				       &cfg->peer, NULL);
+	}
+
+	if (!cfg->data.start && !cfg->data.end) {
```
Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,122 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+		return err;
+	}
+
+	BT_DBG("stored SC for %s (%s, %u-%u)", bt_addr_le_str(&cfg->peer),
```
Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,122 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+}
+
+#if defined(CONFIG_BT_SETTINGS)
+static int sc_store(struct gatt_sc_cfg *cfg)
```
Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1740,24 +1893,30 @@ u8_t bt_gatt_check_perm(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+		return;
+	}
+
+	if (!cfg->data.start && !cfg->data.end) {
```
Done in latest push!

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3904,6 +4166,14 @@ static int db_hash_commit(void)
+	/**
+	 * GATT database has been modified since last boot, likely due to
+	 * a firmware update or a dynamic service that was not re-registered on
+	 * boot. Indicate Service Changed for the full database range to
```
Done in latest push!

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,132 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+	return true;
 }
 
+struct _bt_gatt_ccc sc_ccc = BT_GATT_CCC_INITIALIZER(NULL, sc_ccc_cfg_write,
```
static

fnde-ot replied:
Of course! Just pushed this change.

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -3822,6 +4005,85 @@ static int ccc_set(const char *name, size_t len_rd, settings_read_cb read_cb,
+		BT_DBG("Restored SC for %s", bt_addr_le_str(&addr));
+	} else if (cfg) {
+		/* Clear configuration */
+		cfg->id = BT_ID_DEFAULT;
```
Done in latest push!

fnde-ot commented:
Latest push changes name of "static int bt_gatt_clear_sc" to "static int sc_clear_by_addr".

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1527,11 +1550,11 @@ static u8_t notify_cb(const struct bt_gatt_attr *attr, void *user_data)
-				sc_save(cfg, data->ind_params);
+			if (IS_ENABLED(CONFIG_BT_GATT_SERVICE_CHANGED)) {
+				if (ccc->cfg_changed == sc_ccc_cfg_changed) {
+					sc_save(cfg, data->ind_params);
```
It looks like sc_save is inside a #if defined(CONFIG_BT_GATT_SERVICE_CHANGED) block so Im not sure we can replace like this.

fnde-ot replied:
Yes, just fixed it in latest push alongside another similar place.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -654,34 +771,33 @@ static void sc_indicate_rsp(struct bt_conn *conn,
 
 static void sc_process(struct k_work *work)
 {
-	struct gatt_sc *sc = CONTAINER_OF(work, struct gatt_sc, work);
```
It doesn't seems we changed the storage of gatt_sc so I wonder why you are chaning the handling here?

fnde-ot replied:
@joerchan suggested this change a few hours ago and I agreed!
"gatt_sc is a singleton. Should we maybe remove the CONTAINER_OF in this function, and use gatt_sc?"

Vudentz replied:
I prefer the way it was since that would be easier to change if we ever change the gatt_sc works.

fnde-ot replied:
I'm fine with either ways. @joerchan, @jhedberg, any strong opinions on this before I change it back to CONTAINER_OF? :-D

fnde-ot replied:
I actually just reverted to use CONTAINER_OF in latest push, it seems that it is common practice with many workQ jobs and it would allow for more flexibility.

joerchan replied:
That's alright with me.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -786,28 +908,32 @@ static bool update_range(u16_t *start, u16_t *end, u16_t new_start,
 }
 
-static void sc_indicate(struct gatt_sc *sc, u16_t start, u16_t end)
+static void sc_indicate(struct bt_conn *conn, u16_t start, u16_t end)
```
You don't seem to use the conn here? Was that supposed to be passed over to process with gatt_sc.conn?

fnde-ot replied:
Yes it was supposed to be passed to sc_process using gatt_sc.conn, so to send indication indicate to a specific connected peer or to everyone (conn=NULL). Will fix!

fnde-ot replied:
Just added the missing assignment in latest push.

Vudentz replied:
But the changes usually don't originate from a peer, there either for adding or removing services locally which affects all connections.

fnde-ot replied:
Yes, but sc_indicate can also be called when a single peer reconnects (see sc_restore function), and in this case we need to send an indication to this specific connection only.

joerchan replied:
@Vudentz When the GATT database is changed then all connected peers are are indicated i.e sc_indicate(conn = NULL). For all bonded devices an update is stored. When that peer connects there will be an indicate to that peer ,i.e sc_indicate(conn = connected_peer) called from sc_restore.

Vudentz replied:
I though we wouldn't use sc_indicate for that, or gatt_sc, since in that case we want to indicate immediately and not use a delayed_work for that, gatt_sc aggregate all the changes happening in the meantime if we overwrite it with the range from settings that is no longer a valid range so we would have to lock the gatt_sc in the meantime.

Vudentz replied:
Actually this problem already exists in master, since sc_restore can be called multiple times in a row merging the ranges for different peers, we should probably change it to just operate in the connection passed to ccc_update and sending the indication immediately.

fnde-ot replied:
I just pushed a solution where sc_restore uses bt_gatt_indicate directly instead of sc_indicate. It adds an sc_restore_params array (1 per connection) as well as an sc_restore_rsp callback to handle the change-aware bit in CF.
Seems to work well AFAICanTest.

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -861,7 +987,7 @@ int bt_gatt_service_register(struct bt_gatt_service *svc)
 	}
 
-	sc_indicate(&gatt_sc, svc->attrs[0].handle,
+	sc_indicate(NULL, svc->attrs[0].handle,
```
Not sure what is your intend with passing NULL here?

fnde-ot replied:
Passing conn=NULL means to send an indication to all subscribed devices (or store if those are not connected).

Vudentz commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -877,7 +1003,7 @@ int bt_gatt_service_unregister(struct bt_gatt_service *svc)
 	}
 
-	sc_indicate(&gatt_sc, svc->attrs[0].handle,
+	sc_indicate(NULL, svc->attrs[0].handle,
```
Ditto.

fnde-ot replied:
See above.

fnde-ot commented:
Latest push is just a rebase to make sure it is mergeable.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -1087,12 +1216,48 @@ static u8_t gatt_foreach_iter(const struct bt_gatt_attr *attr,
+						&num_matches,
+						func, user_data) ==
+				BT_GATT_ITER_STOP) {
+				return;
```
Something is not right with the indentation here. BT_GATT_ITER_STOP shouldn't have the same indentation as the return. The split parameters in the gatt_foreach_iter call also seem to be wrong. You should be using tabs + 0-7 spaces to align with the right opening parenthesis of the first line.

fnde-ot replied:
Fixed in latest push.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,127 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+	return NULL;
+}
+
+#if defined(CONFIG_BT_SETTINGS)
```
Is this #if really needed? What you you drop it and use IS_ENABLED in the places where you need to call this function?

fnde-ot replied:
sc_store(...) uses bt_settings_encode_key(...) and settings_save_one(...), which are only compiled in when CONFIG_BT_SETTINGS and CONFIG_SETTINGS are set.

If you prefer I could however put the #if inside the function (as in sc_clear) and use IS_ENABLED when calling it.

jhedberg replied:
What you describe sounds exactly like the use case that IS_ENABLED exists for. As long as the code can see the prototypes of the functions in question in their header files it doesn't matter if they're part of the final image or not, since the calls to them don't exist anymore at the linking stage (the compiler will have stripped them out)

fnde-ot replied:
Yes that's right, will test this right away! Sorry I'm not used to those IS_ENABLED thingies.

fnde-ot replied:
Alright, latest push addresses this and makes more use of IS_ENABLED now that I know the tricks. I tested that it builds without Warnings with multiple config scenarios and it seems to work just fine and not affect the image size at all!

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,127 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+};
+
+#define SC_CFG_MAX (CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)
+static struct gatt_sc_cfg sc_cfg[SC_CFG_MAX] = {};
```
The initialisation to {} is unnecessary since this is a static variable.

fnde-ot commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,127 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
-#if defined(CONFIG_BT_GATT_DYNAMIC_DB)
-static void sc_ccc_cfg_changed(const struct bt_gatt_attr *attr,
-			       u16_t value)
+#if defined(CONFIG_BT_GATT_SERVICE_CHANGED)
```
Fixed in latest push, also in multiple other places (see above comment).

Vudentz reviewed (approved):
LGTM

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,125 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+};
+
+#define SC_CFG_MAX (CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)
+static struct gatt_sc_cfg sc_cfg[SC_CFG_MAX];
```
Should we maybe just use BT_GATT_CCC_MAX here instead?

fnde-ot replied:
In the future we might have CCC loaded on demand when a device connects and BT_GATT_CCC_MAX could then be modified to only be equal to BT_MAX_CONN, so I think it is preferable to leave it as a separate define for now.

joerchan commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -180,13 +183,125 @@ BT_GATT_SERVICE_DEFINE(_2_gap_svc,
+} __packed;
+
+struct gatt_sc_cfg {
+	u8_t		id;
```
id and peer addr is duplicated in gatt_sc_cfg and sc_ccc.cfg. Could we maybe just have an array of `struct sc_data`, and use the same index to save some context data?

fnde-ot replied:
If you mean we should add a "struct sc_data" to bt_gatt_ccc_cfg, the issue is that we add (4 * BT_GATT_CCC_MAX) bytes in every defined CCC attribute and not only the one used for SC, which can add up very quickly. The drawback of course is that it requires additional lookups when we need to do something with SC data.

Maybe in the future we could use the connection index and bonded device index (in keys array) instead of  having to do lookups using id and address.

joerchan replied:
No, that's not what i meant. I meant to remove those two parameters, and use the same index into both gatt_scf_cfg and sc_ccc.cfg to link the information. Just a thought.

fnde-ot replied:
Ah ok, I see what you meant! It's a nice idea, but if we later have CCC loaded on demand (and I would like to work on that soon), we will only be left with CCC data for connected peers and we will then miss indexes for bonded devices, which we will most likely still need for SC data.
I'll see if something like that can be done once I get to it, thanks for the suggestion!

joerchan reviewed (approved):
Looks good. I have a few questions though that you might want to consider. But will approve anyway.
